File: sorting/python/bubble_sort.py
"""
Course: CS 41600
Professor: Dr. Dai
Name: Fabian Rojas
ID: 13
Date: 02/23/2026
File: bubblesort.py
"""

"""
Description:
        Bubble Sort implementation in Python.
        Module function `bubble_sort(arr)` that performs an in-place bubble sort.

Data Type Support:
        - Sorting is performed using Python comparison operators on the elements
            of the supplied list.
        - Numeric types (integers, `decimal.Decimal`, floats, etc.) are sorted
            numerically.
        - Strings are sorted lexicographically (standard string comparison).
"""


def bubble_sort(arr):
    # Calculate the number of elements once so we don't recompute it.
    n = len(arr)

    # Outer loop: we will make at most 'n' full passes through the list.
    # After each pass the largest remaining unsorted element "bubbles" to
    # its correct position at the end of the array, so we can reduce the
    # inner loop range by 'i' on subsequent passes.
    for i in range(n):
        # Track whether any swaps happened during this pass. If no swaps
        # happen, the list is already sorted and we can exit early.
        swapped = False

        # Inner loop: compare adjacent items and swap if out of order.
        # We stop at 'n - i - 1' because the last 'i' elements are already
        # guaranteed to be in correct position after 'i' passes.
        for j in range(0, n - i - 1):
            # If the left element is greater than the right element,
            # swap them to move the larger element to the right.
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True

        # If no swaps occurred in the entire pass, the list is sorted.
        if not swapped:
            break

    # Return the (now sorted) list. Note that bubble sort operates in-place
    # on the provided list and also returns it for convenience.
    return arr


if __name__ == "__main__":
    import sys
    from decimal import Decimal, InvalidOperation, getcontext

    # Read all input tokens from stdin
    tokens = []
    for line in sys.stdin:
        parts = line.strip().split()
        tokens.extend(parts)

    if not tokens:
        sys.exit(0)

    # Try to parse all tokens as Decimal numbers; if any fail, fall back to strings
    data = []
    max_digits = 0
    for t in tokens:
        digits_only = "".join(ch for ch in t if ch.isdigit())
        if len(digits_only) > max_digits:
            max_digits = len(digits_only)

    try:
        getcontext().prec = max(28, max_digits + 5)
    except Exception:
        pass

    try:
        for t in tokens:
            data.append(Decimal(t))
    except (InvalidOperation, ValueError):
        # Fallback to strings
        data = tokens

    # Sort using bubble_sort
    sorted_data = bubble_sort(list(data))

    # Output one value per line
    for item in sorted_data:
        print(item)

File: sorting/java/BubbleSort.java
/*
   Course: CS 41600
   Professor: Dr. Dai
   Name: Jackson Trader
   ID: 17
   Date: 2/23/2026
   File: BubbleSort.java

   Description:
      This program implements the Bubble Sort algorithm to sort a list of
      integers. It reads integers from standard input (one per line), sorts
      them in ascending order using Bubble Sort with early termination
      optimization, and prints the sorted result to standard output (one
      integer per line). Input/output redirection is used to read from and
      write to files.
*/

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.IOException;
import java.util.ArrayList;

public class BubbleSort {
    /**
     * Reads integers from standard input, one per line.
     * 
     * Uses a BufferedReader for efficient I/O when handling
     * large input files. Blank lines are skipped.
     * 
     * @return an ArrayList of integers read from stdin
     * @throws IOException if an I/O error occurs while reading
     */
    public static ArrayList<Integer> readInput() throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        ArrayList<Integer> numbers = new ArrayList<>();
        String line;

        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (!line.isEmpty()) {
                numbers.add(Integer.parseInt(line));
            }
        }

        return numbers;
    }

    /**
     * Sorts an ArrayList of integers in ascending order using the
     * Bubble Sort algorithm.
     * 
     * Bubble Sort works by repeatedly stepping through the list,
     * comparing adjacent elements, and swapping them if they are
     * in the wrong order. This implementation includes an early
     * termination optimization: if no swaps are made during a
     * complete pass, the list is already sorted and the algorithm
     * stops early.
     * 
     * @param numbers the ArrayList of integers to sort
     */
    public static void bubbleSort(ArrayList<Integer> numbers) {
        int n = numbers.size();

        for (int i = 0; i < n - 1; i++) {
            boolean swapped = false;

            for (int j = 0; j < n - i - 1; j++) {
                if (numbers.get(j) > numbers.get(j + 1)) {
                    // Swap adjacent elements
                    int temp = numbers.get(j);
                    numbers.set(j, numbers.get(j + 1));
                    numbers.set(j + 1, temp);
                    swapped = true;
                }
            }

            // If no swaps occurred, the array is already sorted
            if (!swapped) {
                break;
            }
        }
    }

    /**
     * Prints each integer in the ArrayList to standard output,
     * one per line.
     * 
     * Uses a BufferedWriter for efficient I/O when handling
     * large output.
     * 
     * @param numbers the sorted ArrayList of integers to print
     * @throws IOException if an I/O error occurs while writing
     */
    public static void printOutput(ArrayList<Integer> numbers) throws IOException {
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

        for (int num : numbers) {
            writer.write(Integer.toString(num));
            writer.newLine();
        }

        writer.flush();
    }

    /**
     * Main method that orchestrates the Bubble Sort program.
     * 
     * Reads integers from standard input, sorts them using Bubble
     * Sort, and prints the sorted result to standard output. This
     * program is designed to be used with input/output redirection,
     * for example:
     * java BubbleSort < input.txt > output.txt
     * 
     * @param args command-line arguments (not used)
     */
    public static void main(String[] args) {
        try {
            ArrayList<Integer> numbers = readInput();
            bubbleSort(numbers);
            printOutput(numbers);
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}
